kernel/src/hil/uart.rs

```diff
@@ -44,35 +57,223 @@ pub enum Error {
     /// - EINVAL: Impossible parameters (e.g. a `baud_rate` of 0)
     /// - ENOSUPPORT: The underlying UART cannot satisfy this configuration.
-    fn configure(&self, params: UARTParameters) -> ReturnCode;
+    fn configure(&self, params: Parameters) -> ReturnCode;
+}
+
+pub trait Transmit<'a> {
+    /// Set the transmit client, which will be called when transmissions
+    /// complete.
+    fn set_transmit_client(&self, client: &'a TransmitClient);
+
+    /// Transmit a buffer of data. On completion, `transmitted_buffer`
+    /// in the `TransmitClient` will be called.  If the `ReturnCode`
+    /// of `transmit`'s return tuple is SUCCESS, the `Option` will be
+    /// `None` and the struct will issue a `transmitted_buffer`
+    /// callback in the future. If the value of the `ReturnCode` is
+    /// not SUCCESS, then the `tx_data` argument is returned in the
+    /// `Option`. Other valid `ReturnCode` values are:
+    ///  - EOFF: The underlying hardware is not available, perhaps because
+    ///          because it has not been initialized or in the case of a shared
+    ///          hardware USART controller because it is set up for SPI.
+    ///  - EBUSY: the UART is already transmitting and has not made a
+    ///           transmission callback yet.
+    ///  - ESIZE : `tx_len` is larger than the passed slice.
+    ///  - FAIL: some other error.
+    ///
+    /// Each byte in `tx_data` is a UART transfer word of 8 or fewer
+    /// bits.  The word width is determined by the UART configuration,
+    /// truncating any more significant bits. E.g., 0x18f transmitted in
+    /// 8N1 will be sent as 0x8f and in 7N1 will be sent as 0x0f. Clients
+    /// that need to transfer 9-bit words should use `transmit_word`.
+    ///
+    /// Calling `transmit_word` while there is an outstanding
+    /// `transmit_buffer` or `transmit_word` operation will return EBUSY.
+    fn transmit_buffer(&self, tx_data: &'static mut [u8], tx_len: usize) -> (ReturnCode, Option<&'static mut [u8]>);
+
+
+    /// Transmit a single word of data asynchronously. The word length is
+    /// determined by the UART configuration: it can be 6, 7, 8, or 9 bits long.
+    /// If the `ReturnCode` is SUCCESS, on completion,
+    /// `transmitted_word` will be called on the `TransmitClient`.
+    /// Other valid `ReturnCode` values are:
+    ///  - EOFF: The underlying hardware is not available, perhaps because
+    ///          because it has not been initialized or in the case of a shared
+    ///          hardware USART controller because it is set up for SPI.
+    ///  - EBUSY: the UART is already transmitting and has not made a
+    ///           transmission callback yet.
+    ///  - FAIL: some other error.
+    /// If the `ReturnCode` is not SUCCESS, no callback will be made.
+    /// Calling `transmit_word` while there is an outstanding
+    /// `transmit_buffer` or `transmit_word` operation will return
+    /// EBUSY.
+    fn transmit_word(&self, word: u32) -> ReturnCode;
```

Comment:
I probably missed this somewhere, but what is the motivation for the tx and rx `_word` interfaces?

Reply:
Louis explicitly asked for it, as it's a common "get things working" first step and some protocols want to use it heavily.

https://groups.google.com/forum/?utm_medium=email&utm_source=footer#!msg/tock-dev/tn8GI6GQ_yo/jv5eOLLtCwAJ